src/hotspot/share/code/nmethod.cpp

```diff
@@ -417,6 +425,558 @@ static int adjust_pcs_size(int pcs_size) {
+  if (NativeCall::is_call_before(pc)) {
+    NativeCall* ncall = nativeCall_before(pc);
+    return attached_method(ncall->instruction_address());
+  }
+  return nullptr; // not a call
+}
+
+void nmethod::clear_inline_caches() {
+  assert(SafepointSynchronize::is_at_safepoint(), "clearing of IC's only allowed at safepoint");
+  RelocIterator iter(this);
+  while (iter.next()) {
+    iter.reloc()->clear_inline_cache();
+  }
+}
+
+#ifdef ASSERT
+// Check class_loader is alive for this bit of metadata.
+class CheckClass : public MetadataClosure {
+  void do_metadata(Metadata* md) {
+    Klass* klass = nullptr;
+    if (md->is_klass()) {
+      klass = ((Klass*)md);
+    } else if (md->is_method()) {
+      klass = ((Method*)md)->method_holder();
+    } else if (md->is_methodData()) {
+      klass = ((MethodData*)md)->method()->method_holder();
+    } else {
+      md->print();
+      ShouldNotReachHere();
+    }
+    assert(klass->is_loader_alive(), "must be alive");
+  }
+};
+#endif // ASSERT
+
+
+static void clean_ic_if_metadata_is_dead(CompiledIC *ic) {
+  ic->clean_metadata();
+}
+
+// Clean references to unloaded nmethods at addr from this one, which is not unloaded.
+template <typename CallsiteT>
+static void clean_if_nmethod_is_unloaded(CallsiteT* callsite, nmethod* from,
+                                         bool clean_all) {
+  CodeBlob* cb = CodeCache::find_blob(callsite->destination());
+  if (!cb->is_nmethod()) {
+    return;
+  }
+  nmethod* nm = cb->as_nmethod();
+  if (clean_all || !nm->is_in_use() || nm->is_unloading() || nm->method()->code() != nm) {
+    callsite->set_to_clean();
+  }
+}
+
+// Cleans caches in nmethods that point to either classes that are unloaded
+// or nmethods that are unloaded.
+//
+// Can be called either in parallel by G1 currently or after all
+// nmethods are unloaded.  Return postponed=true in the parallel case for
+// inline caches found that point to nmethods that are not yet visited during
+// the do_unloading walk.
+void nmethod::unload_nmethod_caches(bool unloading_occurred) {
+  ResourceMark rm;
+
+  // Exception cache only needs to be called if unloading occurred
+  if (unloading_occurred) {
+    clean_exception_cache();
+  }
+
+  cleanup_inline_caches_impl(unloading_occurred, false);
+
+#ifdef ASSERT
+  // Check that the metadata embedded in the nmethod is alive
+  CheckClass check_class;
+  metadata_do(&check_class);
+#endif
+}
+
+void nmethod::run_nmethod_entry_barrier() {
+  BarrierSetNMethod* bs_nm = BarrierSet::barrier_set()->barrier_set_nmethod();
+  if (bs_nm != nullptr) {
+    // We want to keep an invariant that nmethods found through iterations of a Thread's
+    // nmethods found in safepoints have gone through an entry barrier and are not armed.
+    // By calling this nmethod entry barrier, it plays along and acts
+    // like any other nmethod found on the stack of a thread (fewer surprises).
+    nmethod* nm = as_nmethod_or_null();
```

Comment:
Calling as_nmethod_or_null() from within functions in the nmethod class is suspicious. Shouldn't all such usages be removed? (I'm fine with doing that as a separate change)

Reply:
Sounds good.